src/glang/lib/std/unicode.c3

```diff
@@ -0,0 +1,412 @@
+    }
+
+    // Byte 4
+    if !self.underlying:has_next() {
+        return invalid_unicode_scalar();
+    }
+
+    let forth_byte : u8 = self.underlying:peak();
+    if !is_trailing_utf8_byte_in_valid_range(forth_byte) {
+        return invalid_unicode_scalar();
+    }
+
+    self.underlying:__builtin_get_next();
+    return unsafe_decode_utf8_bytes({first_byte, second_byte, third_byte, forth_byte});
+}
+
+@implicit has_next : (self : UTF8DecodeIter mut&) -> bool = {
+    return self.underlying:has_next();
+}
+
+function encode_to_utf8 : (scalar : UnicodeScalar, buffer: Span<u8>) -> View<u8> = {
+    if is_in_range(scalar.value, 0x00, 0x7f) {
+        buffer:get(0) = Narrow<u8>(scalar.value & 0x7f);
+        return {&buffer.data, 1};
+    }
+
+    if is_in_range(scalar.value, 0x0080, 0x07ff) {
+        buffer:get(0) = Narrow<u8>(0xc0 | ((scalar.value >> 6) & 0x1f));
+        buffer:get(1) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 2};
+    }
+
+    if is_in_range(scalar.value, 0x0800, 0x0fff)
+        or is_in_range(scalar.value, 0x1000, 0xcfff)
+        or is_in_range(scalar.value, 0xd000, 0xd7ff)
+        or is_in_range(scalar.value, 0xe000, 0xffff)
+    {
+        buffer:get(0) = Narrow<u8>(0xe0 | ((scalar.value >> 12) & 0x0f));
+        buffer:get(1) = Narrow<u8>(0x80 | ((scalar.value >> 6) & 0x3f));
+        buffer:get(2) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 3};
+    }
+
+    if is_in_range(scalar.value, 0x10000, 0x3ffff)
+        or is_in_range(scalar.value, 0x40000, 0xfffff)
+        or is_in_range(scalar.value, 0x100000, 0x10ffff)
+    {
+        buffer:get(0) = Narrow<u8>(0xf0 | ((scalar.value >> 18) & 0x07));
+        buffer:get(1) = Narrow<u8>(0x80 | ((scalar.value >> 12) & 0x3f));
+        buffer:get(2) = Narrow<u8>(0x80 | ((scalar.value >> 6) & 0x3f));
+        buffer:get(3) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 4};
+    }
+
+    // Graphene cannot generate an invalid unicode scalar via `make<>` or the iterators
+    // This is a bug in the user's code so I don't mind the unrecoverable error here
+    runtime_assert(false);
+    return {&buffer.data, 0};
+}
+
+function encode_to_utf8 : (scalar : UnicodeScalar, buffer: u8[4] mut&) -> View<u8> = {
+    return encode_to_utf8(scalar, make<Span<u8>>(&mut buffer));
+}
+
+// --------- UTF16 -----------
+function get_next_codepoint : (self: UTF16DecodeIter mut&) -> Optional<u16> = {
+    if !self.underlying:has_next() { return make<Optional<u16>>(); }
+    let first_byte : u8 = self.underlying:__builtin_get_next();
+
+    if !self.underlying:has_next() { return make<Optional<u16>>(); }
+    let second_byte : u8 = self.underlying:__builtin_get_next();
+
+    if self.is_little_endian {
+        return make<Optional<u16>>(first_byte | (as_u16(second_byte) << Narrow<i16>(8)));
+    } else {
+        return make<Optional<u16>>(second_byte | (as_u16(first_byte) << Narrow<i16>(8)));
+    }
```

Comment:
Should just add a byte swap function in the standard library

Reply:
I think this should be a full API. Opened as an issue.